File: src/lib/llm/error-normalizer.ts
/**
 * 错误分类
 */
export enum ErrorCategory {
  NETWORK = 'network',
  AUTH = 'auth',
  RATE_LIMIT = 'rate_limit',
  INVALID_REQUEST = 'invalid_request',
  SERVER_ERROR = 'server_error',
  QUOTA_EXCEEDED = 'quota_exceeded',
  TIMEOUT = 'timeout',
  UNKNOWN = 'unknown',
}

/**
 * 标准化后的错误
 */
export interface NormalizedError {
  category: ErrorCategory;
  message: string; // 用户友好消息
  technicalMessage: string; // 技术/调试消息
  retryable: boolean;
  retryAfter?: number; // 重试等待时间（秒）
}

/**
 * 错误标准化器
 *
 * 职责：将各种错误（网络、API、超时等）标准化为统一格式
 * 提供用户友好的错误消息和重试建议
 */
export class ErrorNormalizer {
  /**
   * 标准化错误
   *
   * @param error 原始错误对象
   * @param providerType Provider 类型（可选，用于特定错误处理）
   * @returns 标准化后的错误
   */
  static normalize(error: any, providerType?: string): NormalizedError {
    if (error == null) {
      return {
        category: ErrorCategory.UNKNOWN,
        message: '发生未知错误，请重试',
        technicalMessage: 'null or undefined error',
        retryable: true,
      };
    }

    const errorMsg = error.message || error.toString();
    const errorMsgLower = errorMsg.toLowerCase();
    const errorStatus = error.status || error.statusCode;

    // 1. 网络错误
    if (this.isNetworkError(error, errorMsgLower)) {
      return {
        category: ErrorCategory.NETWORK,
        message: '网络连接失败，请检查您的网络设置',
        technicalMessage: errorMsg,
        retryable: true,
      };
    }

    // 2. 鉴权错误
    if (this.isAuthError(error, errorStatus, errorMsgLower)) {
      return {
        category: ErrorCategory.AUTH,
        message: 'API 密钥无效或已过期，请检查设置',
        technicalMessage: `${errorStatus}: ${errorMsg}`,
        retryable: false,
      };
    }

    // 3. 限流错误
    if (this.isRateLimitError(error, errorStatus, errorMsgLower)) {
      const retryAfter = this.extractRetryAfter(error) || 60;
      const waitTime = this.formatWaitTime(retryAfter);
      return {
        category: ErrorCategory.RATE_LIMIT,
        message: `请求过于频繁，请等待 ${waitTime} 后重试`,
        technicalMessage: errorMsg,
        retryable: true,
        retryAfter,
      };
    }

    // 4. 配额超限
    if (this.isQuotaError(errorMsgLower)) {
      return {
        category: ErrorCategory.QUOTA_EXCEEDED,
        message: 'API 配额已用尽，请升级套餐或明日再试',
        technicalMessage: errorMsg,
        retryable: false,
      };
    }

    // 5. 超时错误
    if (this.isTimeoutError(error, errorMsgLower)) {
      return {
        category: ErrorCategory.TIMEOUT,
        message: '请求超时，请重试',
        technicalMessage: errorMsg,
        retryable: true,
      };
    }

    // 6. 请求格式错误（4xx）
    if (errorStatus >= 400 && errorStatus < 500) {
      return {
        category: ErrorCategory.INVALID_REQUEST,
        message: '请求格式错误，请检查输入内容',
        technicalMessage: errorMsg,
        retryable: false,
      };
    }

    // 7. 服务器错误（5xx）
    if (errorStatus >= 500) {
      return {
        category: ErrorCategory.SERVER_ERROR,
        message: 'API 服务暂时不可用，请稍后重试',
        technicalMessage: errorMsg,
        retryable: true,
      };
    }

    // 8. 未知错误
    return {
      category: ErrorCategory.UNKNOWN,
      message: '发生未知错误，请重试',
      technicalMessage: errorMsg,
      retryable: true,
    };
  }

  /**
   * 判断是否为网络错误
   */
  private static isNetworkError(error: any, msg: string): boolean {
    return (
      msg.includes('network') ||
      msg.includes('fetch') ||
      msg.includes('connection') ||
      msg.includes('err_network') ||
      error.name === 'NetworkError' ||
      error.code === 'ENOTFOUND' ||
      error.code === 'ECONNREFUSED'
    );
  }

  /**
   * 判断是否为鉴权错误
   */
  private static isAuthError(error: any, status: number, msg: string): boolean {
    return (
      status === 401 ||
      status === 403 ||
      msg.includes('401') ||
      msg.includes('403') ||
      msg.includes('unauthorized') ||
      msg.includes('forbidden') ||
      msg.includes('authentication') ||
      msg.includes('api key')
    );
  }

  /**
   * 判断是否为限流错误
   */
  private static isRateLimitError(error: any, status: number, msg: string): boolean {
    return (
      status === 429 ||
      msg.includes('429') ||
      msg.includes('rate limit') ||
      msg.includes('too many requests') ||
      msg.includes('throttle')
    );
  }

  /**
   * 判断是否为配额错误
   */
  private static isQuotaError(msg: string): boolean {
    return (
      msg.includes('quota') ||
      msg.includes('limit exceeded') ||
      msg.includes('insufficient_quota') ||
      msg.includes('billing')
    );
  }

  /**
   * 判断是否为超时错误
   */
  private static isTimeoutError(error: any, msg: string): boolean {
    return (
      msg.includes('timeout') ||
      msg.includes('etimedout') ||
      error.name === 'TimeoutError' ||
      error.code === 'ETIMEDOUT'
    );
  }

  /**
   * 提取重试等待时间
   */
  private static extractRetryAfter(error: any): number | undefined {
    // 从 HTTP Header 提取
    const retryAfterHeader =
      error.headers?.['retry-after'] || error.response?.headers?.['retry-after'];
    if (retryAfterHeader) {
      const parsed = parseInt(retryAfterHeader);
      if (!isNaN(parsed)) {
        return parsed;
      }
    }

    // 从错误消息中提取 "retry after XX seconds" 模式
    const retryAfterMatch = error.message?.match(/retry after (\d+) seconds/i);
    if (retryAfterMatch) {
      return parseInt(retryAfterMatch[1]);
    }

    // 从错误消息中提取 "wait XX seconds" 模式（如 "Please wait 43877 seconds before retrying"）
    const waitMatch = error.message?.match(/wait (\d+) seconds/i);
    if (waitMatch) {
      return parseInt(waitMatch[1]);
    }

    // 从响应体中提取（针对429错误的JSON响应）
    if (error.response || error.statusText) {
      const responseText = error.response || error.statusText || '';
      const waitInResponseMatch = responseText.match(/wait (\d+) seconds/i);
      if (waitInResponseMatch) {
        return parseInt(waitInResponseMatch[1]);
      }
    }

    return undefined;
  }

  /**
   * 格式化等待时间为友好显示
   */
  private static formatWaitTime(seconds: number): string {
    if (seconds < 60) {
      return `${seconds} 秒`;
    } else if (seconds < 3600) {
      const minutes = Math.ceil(seconds / 60);
      return `${minutes} 分钟`;
    } else if (seconds < 86400) {
      const hours = Math.ceil(seconds / 3600);
      return `${hours} 小时`;
    } else {
      const days = Math.ceil(seconds / 86400);
      return `${days} 天`;
    }
  }
}

File: native-ui/app/src/main/java/com/promenar/nexara/data/remote/parser/ErrorNormalizer.kt
package com.promenar.nexara.data.remote.parser

import com.promenar.nexara.domain.generation.GenerationFailure
import com.promenar.nexara.domain.generation.GenerationFailureCode

enum class ErrorCategory {
    NETWORK,
    AUTH,
    RATE_LIMIT,
    INVALID_REQUEST,
    SERVER_ERROR,
    QUOTA_EXCEEDED,
    TIMEOUT,
    UNKNOWN
}

/**
 * 协议/网络层错误的分类结果。
 *
 * 自结构化失败契约起，本类**不再产出任何自然语言展示文案**：
 * - [technicalMessage] 仅供诊断（日志/调试），禁止作为 UI 展示 message 传播；
 * - 调用方应使用 [toFailure] 得到稳定的 [GenerationFailure]（code + 类型化参数）。
 */
sealed class NormalizedError {
    abstract val category: ErrorCategory
    abstract val technicalMessage: String
    abstract val retryable: Boolean
    abstract val retryAfter: Int?

    final override fun toString(): String =
        "NormalizedError(category=$category, retryable=$retryable, retryAfter=$retryAfter)"

    data class Network(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.NETWORK
        override val retryable = true
        override val retryAfter: Int? = null
    }

    data class Auth(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.AUTH
        override val retryable = false
        override val retryAfter: Int? = null
    }

    data class RateLimit(
        override val technicalMessage: String,
        override val retryAfter: Int?
    ) : NormalizedError() {
        override val category = ErrorCategory.RATE_LIMIT
        override val retryable = true
    }

    data class InvalidRequest(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.INVALID_REQUEST
        override val retryable = false
        override val retryAfter: Int? = null
    }

    data class ServerError(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.SERVER_ERROR
        override val retryable = true
        override val retryAfter: Int? = null
    }

    data class QuotaExceeded(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.QUOTA_EXCEEDED
        override val retryable = false
        override val retryAfter: Int? = null
    }

    data class Timeout(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.TIMEOUT
        override val retryable = true
        override val retryAfter: Int? = null
    }

    data class Unknown(
        override val technicalMessage: String
    ) : NormalizedError() {
        override val category = ErrorCategory.UNKNOWN
        override val retryable = true
        override val retryAfter: Int? = null
    }

    /**
     * 转为结构化失败。technical 仅诊断，code + 类型化 retryAfter 参数供上层映射。
     */
    fun toFailure(cause: Throwable? = null): GenerationFailure = when (this) {
        is Network -> GenerationFailure(GenerationFailureCode.NETWORK, emptyMap(), technicalMessage, cause)
        is Auth -> GenerationFailure(GenerationFailureCode.AUTH, emptyMap(), technicalMessage, cause)
        is RateLimit -> GenerationFailure.rateLimited(
            retryAfterSeconds = retryAfter,
            technical = technicalMessage,
            cause = cause,
        )
        is InvalidRequest -> GenerationFailure(GenerationFailureCode.INVALID_REQUEST, emptyMap(), technicalMessage, cause)
        is ServerError -> GenerationFailure(GenerationFailureCode.SERVER, emptyMap(), technicalMessage, cause)
        is QuotaExceeded -> GenerationFailure(GenerationFailureCode.QUOTA, emptyMap(), technicalMessage, cause)
        is Timeout -> GenerationFailure(GenerationFailureCode.TIMEOUT, emptyMap(), technicalMessage, cause)
        is Unknown -> GenerationFailure.unknown(technical = technicalMessage, cause = cause)
    }
}

object ErrorNormalizer {

    fun normalize(error: Throwable?, providerType: ProviderType? = null): NormalizedError {
        if (error == null) {
            return NormalizedError.Unknown(technicalMessage = "null error")
        }

        val errorMsg = error.message ?: error.toString()
        val errorMsgLower = errorMsg.lowercase()
        val statusCode = extractStatusCode(error)

        if (isNetworkError(error, errorMsgLower)) {
            return NormalizedError.Network(technicalMessage = errorMsg)
        }

        if (isAuthError(error, statusCode, errorMsgLower)) {
            return NormalizedError.Auth(technicalMessage = "$statusCode: $errorMsg")
        }

        if (isRateLimitError(error, statusCode, errorMsgLower)) {
            val retryAfter = extractRetryAfter(error) ?: 60
            return NormalizedError.RateLimit(
                technicalMessage = errorMsg,
                retryAfter = retryAfter,
            )
        }

        if (isQuotaError(errorMsgLower)) {
            return NormalizedError.QuotaExceeded(technicalMessage = errorMsg)
        }

        if (isTimeoutError(error, errorMsgLower)) {
            return NormalizedError.Timeout(technicalMessage = errorMsg)
        }

        if (statusCode in 400..499) {
            return NormalizedError.InvalidRequest(technicalMessage = errorMsg)
        }

        if (statusCode >= 500) {
            return NormalizedError.ServerError(technicalMessage = errorMsg)
        }

        return NormalizedError.Unknown(technicalMessage = errorMsg)
    }

    private fun extractStatusCode(error: Throwable): Int {
        return when (error) {
            is HttpStatusException -> error.statusCode
            else -> {
                val msg = error.message ?: ""
                val match = Regex("""\b(?:status[_ ]?code?[:\s]*)?(\d{3})\b""").find(msg)
                val code = match?.groupValues?.lastOrNull()?.toIntOrNull()
                if (code != null && code in 100..599) code else 0
            }
        }
    }

    private fun isNetworkError(error: Throwable, msg: String): Boolean {
        return msg.contains("network") ||
            msg.contains("fetch") ||
            msg.contains("connection") ||
            msg.contains("err_network") ||
            msg.contains("econnrefused") ||
            msg.contains("enotfound") ||
            error::class.simpleName?.lowercase()?.contains("network") == true
    }

    private fun isAuthError(error: Throwable, status: Int, msg: String): Boolean {
        return status == 401 ||
            status == 403 ||
            msg.contains("401") ||
            msg.contains("403") ||
            msg.contains("unauthorized") ||
            msg.contains("forbidden") ||
            msg.contains("authentication") ||
            msg.contains("api key")
    }

    private fun isRateLimitError(error: Throwable, status: Int, msg: String): Boolean {
        return status == 429 ||
            msg.contains("429") ||
            msg.contains("rate limit") ||
            msg.contains("too many requests") ||
            msg.contains("throttle")
    }

    private fun isQuotaError(msg: String): Boolean {
        return msg.contains("quota") ||
            msg.contains("limit exceeded") ||
            msg.contains("insufficient_quota") ||
            msg.contains("billing")
    }

    private fun isTimeoutError(error: Throwable, msg: String): Boolean {
        return msg.contains("timeout") ||
            msg.contains("etimedout") ||
            error::class.simpleName?.lowercase()?.contains("timeout") == true
    }

    private fun extractRetryAfter(error: Throwable): Int? {
        val msg = error.message ?: return null

        val retryAfterMatch = Regex("""retry after (\d+) seconds?""", RegexOption.IGNORE_CASE)
            .find(msg)
        if (retryAfterMatch != null) {
            return retryAfterMatch.groupValues[1].toIntOrNull()
        }

        val waitMatch = Regex("""wait (\d+) seconds?""", RegexOption.IGNORE_CASE)
            .find(msg)
        if (waitMatch != null) {
            return waitMatch.groupValues[1].toIntOrNull()
        }

        if (error is RetryAfterHolder) {
            return error.retryAfter
        }

        return null
    }
}

open class HttpStatusException(
    val statusCode: Int,
    override val message: String
) : Exception(message)

interface RetryAfterHolder {
    val retryAfter: Int?
}
